improv_ble_client/improv_ble_client.py

```diff
@@ -0,0 +1,469 @@
+    ) -> Callable[[], None]:
+        """Subscribe to state updates."""
+        _LOGGER.debug("%s: subscribe_state_updates", self.name)
+
+        async def _subscribe_state_updates() -> Callable[[], None]:
+            return self._notification_handlers.subscribe_state(state_callback)
+
+        return await self._execute(_subscribe_state_updates)
+
+    @retry_bluetooth_connection_error(DEFAULT_ATTEMPTS)  # type: ignore[misc]
+    async def _execute(self, procedure: Callable[[], Coroutine[Any, Any, _T]]) -> _T:
+        """Execute a procedure."""
+        if self._procedure_lock.locked():
+            _LOGGER.debug(
+                "%s: Procedure already in progress, waiting for it to complete; "
+                "RSSI: %s",
+                self.name,
+                self.rssi,
+            )
+        async with self._procedure_lock:
+            try:
+                await self._ensure_connected()
+                return await procedure()
+            except asyncio.CancelledError as err:
+                if self._disconnect_reason is None:
+                    raise ImprovError from err
+                if self._disconnect_reason == DisconnectReason.TIMEOUT:
+                    raise Timeout from err
+                if self._disconnect_reason == DisconnectReason.UNEXPECTED:
+                    raise UnexpectedDisconnect from err
+                raise Disconnected(self._disconnect_reason) from err
+            except ImprovError:
+                self._disconnect(DisconnectReason.ERROR)
+                raise
+
+    async def _ensure_connected(self) -> None:
+        """Ensure connection to device is established."""
+        if self._connect_lock.locked():
+            _LOGGER.debug(
+                "%s: Connection already in progress, waiting for it to complete; "
+                "RSSI: %s",
+                self.name,
+                self.rssi,
+            )
+        if self._client and self._client.is_connected:
+            self._reset_disconnect_timer()
+            return
+        async with self._connect_lock:
+            # Check again while holding the lock
+            if self._client and self._client.is_connected:
+                self._reset_disconnect_timer()
+                return
+            _LOGGER.debug("%s: Connecting; RSSI: %s", self.name, self.rssi)
+            client = await establish_connection(
+                BleakClientWithServiceCache,
+                self._ble_device,
+                self.name,
+                self._disconnected,
+                use_services_cache=False,  # True
+                ble_device_callback=lambda: self._ble_device,
+            )
+            _LOGGER.debug("%s: Connected; RSSI: %s", self.name, self.rssi)
+
+            self._client = client
+            self._disconnect_reason = None
+            self._reset_disconnect_timer()
+
+            _LOGGER.debug(
+                "%s: Subscribe to notifications; RSSI: %s", self.name, self.rssi
+            )
```

Comment:
Maybe verify all the char uuids are present and if not try clearing the cache and reconnecting to the services get resolved again.

Reply:
Can you link to an example doing that?

Reply:
https://github.com/Danielhiversen/pySwitchbot/blob/b460de69cef5629fed7fbb2b81464c6383c2da79/switchbot/devices/device.py#L285

Reply:
I see, that's more or less a workaround for a missing feature in bleak IMHO, and the same implementation is duplicated in multiple BLE-packages.

Maybe it could be moved to its own package, `bleak_utils` or `bleak_helpers`?
I'll copy the implementation to `py-improv-ble` for now though.

Reply:
I'm happy to accept a PR to `bleak_retry_connector`